tools/my-pass/MyPass.cpp

```diff
@@ -0,0 +1,104 @@
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/InstrTypes.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Operator.h"
+#include "llvm/Pass.h"
+
+#include <bit>
+#include <optional>
+#include <tuple>
+
+using namespace llvm;
+
+namespace {
+
+static std::optional<uint64_t> matchConstantPowerOfTwo(const Value *const V) {
+  auto *const C = dyn_cast<ConstantInt>(V);
+  if (C == nullptr) {
+    return std::nullopt;
+  }
+  if (C->getBitWidth() > 64) {
+    return std::nullopt;
+  }
+  if (C->isNegative()) {
+    return std::nullopt;
+  }
+  uint64_t const Int = C->getZExtValue();
+  if (!std::has_single_bit(Int)) {
+    return std::nullopt;
+  }
+  return Int;
+}
+
+using MatchTuple = std::tuple<Value*, uint64_t>;
+
+// Matches a multiplication by a constant power of two.
+// Returns a potentially non-constant multiplicand and a constant multiplier.
+static std::optional<MatchTuple> matchOperands(const MulOperator *const O) {
+  Value *const LHS = O->getOperand(0);
+  Value *const RHS = O->getOperand(1);
+
+  if (const auto Match = matchConstantPowerOfTwo(RHS)) {
+    return MatchTuple{ LHS, *Match };
+  }
+
+  if (const auto Match = matchConstantPowerOfTwo(LHS)) {
+    return MatchTuple{ RHS, *Match };
+  }
+
+  return std::nullopt;
+}
```

Comment:
This is correct, but it's a bit reinventing stuff that LLVM already provides here: https://github.com/revng/llvm-project/blob/master/llvm/include/llvm/IR/PatternMatch.h

If you want to learn the LLVM way you can try and use that instead of rolling your stuff.